docs/proposals/improve-addons.md

```diff
@@ -80,230 +84,176 @@ See below for a high level architecture view. Each component will be explained i
+  values: # Both structure and values depend on what is configured in the application
     abc: xyz
 ```
 
-### CoreAddonController
-
-The CoreAddonController resides in the seed cluster. Its main purpose is to manage AddonInstallation CRs for CoreAddons. This job entails standard create/update/delete operations as well as picking the correct CoreAddons per cluster. The picking is required as some CoreAddons are specific to the type of cluster (e.g. chosen cloud provider, chosen CNI,...). Additionally the CoreAddonController protects against accidental deletion of AddonInstallation CRs of type Core. We think this makes sense as losing a CoreAddon (e.g. OpenVPN) could be detrimental to User Cluster health.
-
-In order to know which UserCluster requires which CoreAddons, the CoreAddonController  works by maintaining a sub-list of CoreAddons for each UserCluster computed from the following sources:
-
-- cluster.kubermatic.io/v1 CR → Will be used for CoreAddons that depend on a clusters configuration (e.g. cilium vs canal). The logic for this will be coded directly into CoreAddonController as it can become quite complicated
-- CustomCoreAddon configmap →A configmap that contains a list of addons that should be added to every cluster. This map can be edited by the Cluster Admin. This allows Cluster Admins to specify their own Addons that should be shipped with every cluster, which is in line with the current [defaultManifests](https://docs.kubermatic.com/kubermatic/v2.18/guides/addons/) functionality
-
-In order to protect against accidental deletion of an AddonInstallation of type Core, we propose to make use of k8s [finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). In this case the CoreAddonController would automatically add itself as a finalizer to any AddonInstallation it creates. Then whenever an AddonInstallation is being deleted, kubernetes will wait with the deletion until the finalizer is removed. Afterwards CoreAddonController is going to run a couple of checks internally:
+### ApplicationInstallationController
 
-1. Ask KKP api if the cluster was marked for deletion → if yes, remove finalizer and internal addon list; We need this check to ensure that CoreAddonController is not blocking the deletion of a cluster
-2. Check if the CoreAddon is still in internal list
-   - yes →keep CR as is to block deletion
-   - no →remove finalizer; CR will be deleted by k8s. This allows ClusterAdmins to remove CoreAddons if they really want to by removing them from CoreAddonControllers internal list, but still protects against accidental deletion
+The ApplicationInstallationController resides in the seed cluster. Its main purpose is to watch ApplicationInstallation CRs and ensure that the requested Applications are installed/modified/deleted in the corresponding user clusters. By extension of this, the ApplicationInstallationController needs to be able to fetch the manifests for Applications.
```

Comment:
It makes sense if we want to expose ApplicationInstallation CR to the end user, in the user-cluster API. A-la MachineDeployment. But the question is: is it the case?

Reply:
I think it is, in my opinion this should be exposed to the end user (user cluster admin). @SimonTheLeg @vgramer what's your point of view on this?

Reply:
and @shaase-ctrl ^

Reply:
Sorry, I did not make it in time over the weekend. 
We need to have a user cluster admin be able to install apps for his cluster. KKP admins that need to configure everything per cluster is contrary to the self-service platform we built.

Reply:
I don't think it's possible to expose it, as it's defined as

> contains a mapping of a specific Application and cluster where to install

Reply:
With the current design, KKP admin defined possible applications to install by creating an `ApplicationDefinition` CR in the master cluster. The `ApplicationDistributor` controller distributes these CRs to all seeds. 

the `ApplicationInstallation` CR  contains a reference to the `ApplicationDefinition` CR and the version of the application. this reference is used to know how to install the application.

if we move the `ApplicationInstallation` to the user-cluster and the `ApplicationInstallationController` into the cluster-namespace. That means we have to either:

1.  have a client to the seed cluster in the `ApplicationInstallationController` to be able to retrieve to the corresponding `ApplicationDefinition` CR. 
2. have a controller (ApplicationDistributor controller ?) that replicates `ApplicationDefinition` CR into each user cluster. So the `ApplicationInstallationController` can retrieve it.


Solution 1 seems pretty messy to me. especially if we consider deletion of an `ApplicationDefinition` CR  by a KKP admin. what happens to applications already installed.  

With solution 2, the deletion can be handled by an ownerRef set to `ApplicationInstalltion` CR inside the user-cluster.  When a KKP admin deletes an `ApplicationDefinition` CR, the deletion is eventually propagated to user-cluster and applications are removed.

Reply:
In addition to @vgramer comments, maybe a short background and two follow-up questions from my side

Background
We had two initial ideas behind placing the ApplicationInstallationCRs inside the seed:
1. It would be tough for cluster admins to create invalid ApplicationInstallationCRs, as they can only use the UI to create them. This was born out of the idea that the majority of Application users would use the UI. Which sounds plausible to me, because if someone is more proficient in k8s and knows how to use helm, kustomized, etc., they would not really need the KKP Application feature. => now I think we can drop this issue, as we extend our ValidatingWebhook to ensure that no invalid ApplicationInstallationCRs can be created/edited (I think we just need to add that users cannot change the status field, otherwise Application state could get messy). @shaase-ctrl from a product-mgmt perspective: should the Cluster admin be able to edit ApplicationInstallations using the k8s api or would it be sufficient if they can do it via the UI?
2. Reduce the number of cross-cluster calls. We need to be aware of the fact that moving the ApplicationInstallationCRs into the user clusters will result in more cross-cluster calls. This is due to the fact that we need to keep ApplicationDefinitions in the seed cluster (otherwise cluster admins could just edit the definitions and make up their own Applications). This in turn means that for almost every interaction with the ApplicationInstallation we need to call the seed. From my previous experience I would always try to reduce cross cluster calls, but I am not sure how we feel about them at Kubermatic. Who would be the best person to weigh in on this?

Reply:
Using the seeds for such features is the intention of how I would like to evolve KKP in general. We deliver additional features via the seeds in order to steer traffic, access etc. 

1. UI and API, it is as you reasoned, in the beginning more UI, later on more automation via API. For an MVP we could start with UI and go with API later (but it should be possible). The spread of the user base is quite large in the end.

2. Very good point for the cross cluster calls, I do not think that we optimized for this, but it is a good consideration. Overall this is a clash between control for KKP admin and flexibility for user cluster admin. My thoughts on this revolve around the possibility to use helm anyway in a user cluster for the user cluster admin, but helm could be delivered via the "applications" feature. Control and flexibility if you will. If I understood correctly what @vgramer  said, kkp admins could involuntarily delete apps that are in use, which could be quite unfortunate. In order to get to a result, I think we should go with the kkp admin approach you already have for "applications" and see if there is a huge demand for independently manage "applications" via the user cluster admin. Thanks @mfranczy @SimonTheLeg @vgramer and @kron4eg for your considerations.

Reply:
@SimonTheLeg 

> It would be tough for cluster admins to create invalid ApplicationInstallationCRs

We should have a proper validation (on the client and server side), it should not matter if this is UI or API.

>  because if someone is more proficient in k8s and knows how to use helm, kustomized, etc., they would not really need the KKP Application feature. => now

Initially when I thought about this application mechanism I didn't mean to do it because some users don't know helm, it was for sake of automation and improved user experience (without forcing people to create their own automation on top - that's important with massive cluster creation)

> Reduce the number of cross-cluster calls.

Correct me if I am wrong, but IMO you won't reduce them here in both scenarios:
* the current design - call to update `ApplicationInstallation` status from seed to user cluster.
* when moving `ApplicationInstallation` to user cluster - call to check `ApplicationDefinition`, again seed to user cluster.

@vgramer

> have a client to the seed cluster in the ApplicationInstallationController to be able to retrieve to the corresponding ApplicationDefinition CR.

ApplicationInstallationController has to be aware of seed and user clusters anyway. Even with the current design while keeping the `ApplicationInstallation` CR on the seed you have to make a call to the user cluster to create a resource there over helm, k8s client or whatever source type you support.

Since the controller would run on the seed you have to get a client to user cluster (similar way how the current addon mechanism works).

If you move `ApplicationInstallation` CR to user you could have a custom informer to catch `ApplicationDefinition` events and map them to `ApplicationInstallation` CR events.


After all that's just my opinion, if you still feel different about that and have majority in this go ahead and continue with your idea.